src/nix/narinfo.rs

```diff
@@ -0,0 +1,246 @@
+                    return Ok(Some(Self::parse(&narinfo)?));
+                } else {
+                    log::warn!("Cache responded with error code: {}", response.status());
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Parses a `narinfo` string into a `NarInfo` struct.
+    ///
+    /// This function takes a `narinfo` string and parses it into a `NarInfo` struct.
+    /// It iterates over each line in the `narinfo` string, splitting it into a key-value pair.
+    /// The key is then matched to a corresponding field in the `NarInfo` struct.
+    /// If a key does not match any field, an error is returned.
+    /// If a required field is missing from the `narinfo` string, an error is returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `narinfo` - A string slice that holds the `narinfo` data.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<NarInfo, Error>` - Returns a `NarInfo` struct if parsing is successful, otherwise returns an `Error`.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    /// * A line in the `narinfo` string does not contain a ':' delimiter.
+    /// * The `narinfo` string contains an unknown key.
+    /// * A required field is missing from the `narinfo` string.
+    pub fn parse(narinfo: &str) -> crate::error::Result<Self> {
+        let mut store_path = None;
+        let mut url = None;
+        let mut compression = None;
+        let mut file_hash = None;
+        let mut file_size = None;
+        let mut nar_hash = None;
+        let mut nar_size = None;
+        let mut deriver = None;
+        let mut system = None;
+        let mut references = Vec::new();
+        let mut sig = None;
+        let mut ca = None;
+
+        for line in narinfo.lines() {
+            let (key, value) = line
+                .split_once(':')
+                .ok_or_else(|| crate::error::Error::NarInfoNoDelimiter(line.to_string()))
+                .map(|(key, value)| (key.trim(), value.trim()))?;
+
+            match key {
+                "StorePath" => store_path = Some(value.to_string()),
+                "URL" => url = Some(value.to_string()),
+                "Compression" => compression = Some(value.to_string()),
+                "FileHash" => file_hash = Some(value.to_string()),
+                "FileSize" => file_size = Some(value.parse()?),
+                "NarHash" => nar_hash = Some(value.to_string()),
+                "NarSize" => nar_size = Some(value.parse()?),
+                "Deriver" => deriver = Some(value.to_string()),
+                "System" => system = Some(value.to_string()),
+                "References" => references = value.split(' ').map(|s| s.to_string()).collect(),
+                "Sig" => sig = Some(value.to_string()),
+                "CA" => ca = Some(value.to_string()),
+                _ => return Err(crate::error::Error::NarInfoUnknownKey(key.to_string())),
```

Comment:
Probably don't want to fail here for forward-compat (and because Nix allows extra fields, so there might be in the wild)